{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 362,
   "metadata": {},
   "outputs": [],
   "source": [
    "#!/usr/bin/python\n",
    "\n",
    "import re\n",
    "import cairo\n",
    "import random\n",
    "import argparse\n",
    "\n",
    "parser = argparse.ArgumentParser(description='Given a Fasta file with uppercase exons and a text file with one motif per line, creates an svg image depicting each sequence intronic and exonic region with matching motif regions. NOTE: Uses random color generation, if two motifs share colors that are too similar, please rerun script.')\n",
    "parser.add_argument('-f', '--fasta', metavar='Fasta File Path', required=True,type=str, help='Absolute file path to Fasta file')\n",
    "parser.add_argument('-m', '--motif', metavar='Motif File Path', required=True,type=str, help='Absolute file path to motif file')\n",
    "parser.add_argument('-n', '--name', metavar='Output Image Name', type=str, default=False, help='Optional: Provides the name for the output svg image file')\n",
    "\n",
    "parser = parser.parse_args()\n",
    "\n",
    "if parser.name is False: #Add our conditional to check if a custom name was passed\n",
    "    FileName = \"Sequence_Motif\"\n",
    "else:\n",
    "    FileName = parser.name\n",
    "\n",
    "#Set our passed variables\n",
    "InputFile = parser.fasta\n",
    "InputMotifFile = parser.motif\n",
    "\n",
    "############################\n",
    "#Define our functions\n",
    "\n",
    "#Fixes any ambiguous bases in our motifs for use in your regular expression\n",
    "def motif_fixer (motif):\n",
    "    motif = motif.upper()\n",
    "    motif = motif.replace('Y','[CT]')\n",
    "    motif = motif.replace('R','[AU]')\n",
    "    motif = motif.replace('S','[GC]')\n",
    "    motif = motif.replace('W','[AT]')\n",
    "    motif = motif.replace('K','[GT]')\n",
    "    motif = motif.replace('M','[AC]')\n",
    "    motif = motif.replace('B','[CGT]')\n",
    "    motif = motif.replace('D','[AGT]')\n",
    "    motif = motif.replace('H','[ACT]')\n",
    "    motif = motif.replace('V','[ACG]')\n",
    "    motif = motif.replace('N','.')\n",
    "    return motif\n",
    "\n",
    "#Draws our exons and motifs, depending on start/end sites\n",
    "def draw_line(x,y,start,stop):\n",
    "    context.move_to(x+start,y)\n",
    "    context.line_to(x+stop,y)\n",
    "    context.stroke()\n",
    "    \n",
    "#Draws our legend\n",
    "def draw_legend(r,g,b,n,motif):\n",
    "    context.set_source_rgb(r,g,b)\n",
    "    context.rectangle(n,10,10,10)\n",
    "    context.fill()\n",
    "    context.set_source_rgb(0,0,0)\n",
    "    context.move_to(n + 11,17)\n",
    "    context.show_text(motif)\n",
    "    n = n + 10*(len(motif))\n",
    "    return(n)\n",
    "\n",
    "#Exonseq looks for the exons by searching for capital letters within the sequence looking for uppercase\n",
    "#letters that are grouped together, not a function but a compiled reg-expression search\n",
    "exonseq = re.compile(\"[A-Z]{1,}\")\n",
    "\n",
    "############################\n",
    "#If our fasta file is multi-line, we now convert it into a single header + sequence line file\n",
    "\n",
    "#First we find the total number of lines in the file to account for the final line\n",
    "total_lines = 0\n",
    "fh = open(InputFile)\n",
    "for lines in fh:\n",
    "    total_lines+=1\n",
    "fh.close()\n",
    "\n",
    "#We iterate through our file and join all sequence lines, also makes our file into an iterable list\n",
    "fh = open(InputFile)\n",
    "sequence = \"\"\n",
    "InputFasta = []\n",
    "n=0\n",
    "for lines in fh:\n",
    "    lines = lines.rstrip()\n",
    "    n+=1\n",
    "    if n == 1: #Exception for first line\n",
    "        InputFasta.append(lines)\n",
    "        sequence = \"\"\n",
    "    elif n == total_lines: #Exception for last line\n",
    "        sequence = sequence + lines\n",
    "        InputFasta.append(sequence)\n",
    "    elif lines.startswith(\">\"):\n",
    "        InputFasta.append(sequence)\n",
    "        InputFasta.append(lines)\n",
    "        sequence = \"\"\n",
    "    elif not lines.startswith(\">\"):\n",
    "        sequence = sequence + lines\n",
    "fh.close()\n",
    "\n",
    "############################\n",
    "#Go through both of our input motif and fasta file in order to see how many genes/motifs we'll be looking at in total\n",
    "#so we can create our plot with sufficient room and colors. Additionally create a list of our fixed motifs(fix ambiguousness)\n",
    "\n",
    "#Open our fasta file to count the number of genes to find the longest present and account for plot Y axis\n",
    "entrylength = 0\n",
    "nlines = 0\n",
    "for lines in InputFasta:\n",
    "    if not lines.startswith(\">\"):\n",
    "        nlines+=1\n",
    "        if len(lines) > entrylength:\n",
    "            entrylength = len(lines)\n",
    "\n",
    "#Open our motif file and create a list with fixed (no ambiguous bases) motifs\n",
    "InputMotif = open(InputMotifFile)\n",
    "motifs = []\n",
    "for motif in InputMotif:\n",
    "    motif = motif.rstrip()\n",
    "    motifs.append(motif_fixer(motif))\n",
    "InputMotif.close()\n",
    "\n",
    "#Create three lists pertaining to the colors that will be used to represent each individual motif\n",
    "red = []\n",
    "green = []\n",
    "blue = []\n",
    "for motif in motifs:\n",
    "    red.append(random.randint(0,100)/100)\n",
    "    green.append(random.randint(0,100)/100)\n",
    "    blue.append(random.randint(0,100)/100)\n",
    "    \n",
    "############################\n",
    "#Create our base x and y draw values and initiate our cairo surface by the calculated values\n",
    "x = 25 #This can be used to tune how far fromt he margin we wish to start our images/writing\n",
    "y = 0\n",
    "surface = cairo.SVGSurface(FileName+\".svg\", entrylength+(x*2), (nlines*50)+50)\n",
    "context = cairo.Context(surface)\n",
    "context.set_font_size(8)\n",
    "\n",
    "#Draw a legend of what each motif color means\n",
    "n = 25 #Decides how far down the y acis we wish to start our legend\n",
    "for motif,r,g,b in zip(motifs,red,green,blue):\n",
    "    if n == 25:\n",
    "        n = draw_legend(r,g,b,n,motif)\n",
    "    else:\n",
    "        n = draw_legend(r,g,b,n,motif)\n",
    "\n",
    "############################\n",
    "#The way we will draw the exons will be to first draw an entire line to signify the whole gene (introns) then for each position in which\n",
    "#an upper case character was found (exon), we will draw a vertical line. Then we will look at the motifs and draw different\n",
    "#colored horizontal lines through each respective site\n",
    "\n",
    "for lines in InputFasta:\n",
    "    if lines.startswith(\">\"): #Begin by adding the header above each gene that will be ploted\n",
    "        y = y + 50\n",
    "        context.set_source_rgb(0,0,0)\n",
    "        context.move_to(x,y-15)\n",
    "        context.show_text(lines)\n",
    "        \n",
    "    if not lines.startswith(\">\"): #Draws our 'intron' line which is simply a line the length of our sequence\n",
    "        context.set_line_width(1)\n",
    "        context.move_to(x,y)\n",
    "        context.line_to(x+len(lines),y)\n",
    "        context.stroke()\n",
    "        \n",
    "        #We set our line width to 10 since we will draw our motifs and exons much wider\n",
    "        context.set_line_width(10)\n",
    "        \n",
    "        #Set and reset our exon start and end variables for each sequence\n",
    "        exonstart = []\n",
    "        exonend = []\n",
    "        \n",
    "        #Look through our current line and save all the exon start and stop locations\n",
    "        for bases in exonseq.finditer(lines): #Save the location of every uppercase character in the sequence\n",
    "            exonstart.append(bases.start()+1) #In order to account for 0 based python counts, we used start + 1\n",
    "            exonend.append(bases.end())\n",
    "            \n",
    "        for start, stop in zip(exonstart,exonend): #Iterate through our start and stop sites to draw our exons\n",
    "                draw_line(x,y,start,stop)\n",
    "        \n",
    "        #Additionall we change our sequence lines to uppercase as the regular expressions are looking for uppercase characters\n",
    "        lines = lines.upper()\n",
    "        \n",
    "        #Iterate through our motif list, as well as the corresponding colors for each motif\n",
    "        for motif,r,g,b in zip(motifs,red,green,blue):\n",
    "            #Set and reset our motif stard and end locations for each motif\n",
    "            motifstart = []\n",
    "            motifend = []\n",
    "            motifseq = re.compile(motif,) #Compile our reg-ex search to match for each specific motif\n",
    "        \n",
    "            for bases in motifseq.finditer(lines):\n",
    "                motifstart.append(bases.start()+1) #In order to account for 0 based python counts, we used start + 1\n",
    "                motifend.append(bases.end())\n",
    "        \n",
    "            for start, stop in zip(motifstart,motifend): #Iterate through our start and stop sites to draw our motifs        \n",
    "                context.set_source_rgb(r,g,b)\n",
    "                draw_line(x,y,start,stop)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 264,
   "metadata": {},
   "outputs": [],
   "source": [
    "# #This is the code without the argparse for debugging purposes\n",
    "\n",
    "# #!/usr/bin/python\n",
    "\n",
    "# import re\n",
    "# import cairo\n",
    "# import random\n",
    "\n",
    "# InputFile = \"sequence.txt\"\n",
    "# InputMotifFile = \"motifs2.txt\"\n",
    "# FileName = \"FileName\"\n",
    "\n",
    "# ############################\n",
    "# #Define our functions\n",
    "\n",
    "# #Fixes any ambiguous bases in our motifs for use in your regular expression\n",
    "# def motif_fixer (motif):\n",
    "#     motif = motif.upper()\n",
    "#     motif = motif.replace('Y','[CT]')\n",
    "#     motif = motif.replace('R','[AU]')\n",
    "#     motif = motif.replace('S','[GC]')\n",
    "#     motif = motif.replace('W','[AT]')\n",
    "#     motif = motif.replace('K','[GT]')\n",
    "#     motif = motif.replace('M','[AC]')\n",
    "#     motif = motif.replace('B','[CGT]')\n",
    "#     motif = motif.replace('D','[AGT]')\n",
    "#     motif = motif.replace('H','[ACT]')\n",
    "#     motif = motif.replace('V','[ACG]')\n",
    "#     motif = motif.replace('N','.')\n",
    "#     return motif\n",
    "\n",
    "# #Draws our exons and motifs, depending on start/end sites\n",
    "# def draw_line(x,y,start,stop):\n",
    "#     context.move_to(x+start,y)\n",
    "#     context.line_to(x+stop,y)\n",
    "#     context.stroke()\n",
    "    \n",
    "# #Draws our legend\n",
    "# def draw_legend(r,g,b,n,motif):\n",
    "#     context.set_source_rgb(r,g,b)\n",
    "#     context.rectangle(n,10,10,10)\n",
    "#     context.fill()\n",
    "#     context.set_source_rgb(0,0,0)\n",
    "#     context.move_to(n + 11,17)\n",
    "#     context.show_text(motif)\n",
    "#     n = n + 10*(len(motif))\n",
    "#     return(n)\n",
    "\n",
    "# ############################\n",
    "# #If our fasta file is multi-line, we now convert it into a single header + sequence line file\n",
    "\n",
    "# #First we find the total number of lines in the file to account for the final line\n",
    "# total_lines = 0\n",
    "# fh = open(InputFile)\n",
    "# for lines in fh:\n",
    "#     total_lines+=1\n",
    "# fh.close()\n",
    "\n",
    "# #We iterate through our file and join all sequence lines, also makes our file into an iterable list\n",
    "# fh = open(InputFile)\n",
    "# sequence = \"\"\n",
    "# InputFasta = []\n",
    "# n=0\n",
    "# for lines in fh:\n",
    "#     lines = lines.rstrip()\n",
    "#     n+=1\n",
    "#     if n == 1: #Exception for first line\n",
    "#         InputFasta.append(lines)\n",
    "#         sequence = \"\"\n",
    "#     elif n == total_lines: #Exception for last line\n",
    "#         sequence = sequence + lines\n",
    "#         InputFasta.append(sequence)\n",
    "#     elif lines.startswith(\">\"):\n",
    "#         InputFasta.append(sequence)\n",
    "#         InputFasta.append(lines)\n",
    "#         sequence = \"\"\n",
    "#     elif not lines.startswith(\">\"):\n",
    "#         sequence = sequence + lines\n",
    "# fh.close()\n",
    "\n",
    "# ############################\n",
    "# #Go through both of our input motif and fasta file in order to see how many genes/motifs we'll be looking at in total\n",
    "# #so we can create our plot with sufficient room and colors. Additionally create a list of our fixed motifs\n",
    "\n",
    "# #Open our fasta file to count the number of genes to find the longest present and account for plot Y axis\n",
    "# entrylength = 0\n",
    "# nlines = 0\n",
    "# for lines in InputFasta:\n",
    "#     if not lines.startswith(\">\"):\n",
    "#         nlines+=1\n",
    "#         if len(lines) > entrylength:\n",
    "#             entrylength = len(lines)\n",
    "\n",
    "# #Open our motif file and create a list with fixed (no ambiguous bases) motifs\n",
    "# InputMotif = open(InputMotifFile)\n",
    "# motifs = []\n",
    "# for motif in InputMotif:\n",
    "#     motif = motif.rstrip()\n",
    "#     motifs.append(motif_fixer(motif))\n",
    "# InputMotif.close()\n",
    "\n",
    "# #Create three lists pertaining to the colors that will be used to represent each individual motif\n",
    "# red = []\n",
    "# green = []\n",
    "# blue = []\n",
    "# for motif in motifs:\n",
    "#     red.append(random.randint(0,100)/100)\n",
    "#     green.append(random.randint(0,100)/100)\n",
    "#     blue.append(random.randint(0,100)/100)\n",
    "    \n",
    "# ############################\n",
    "# #Create our base x and y draw values and initiate our cairo surface by the calculated values\n",
    "# x = 25\n",
    "# y = 0\n",
    "# surface = cairo.SVGSurface(FileName+\".svg\", entrylength+(x*2), (nlines*50)+50)\n",
    "# context = cairo.Context(surface)\n",
    "# context.set_font_size(8)\n",
    "\n",
    "# #Exonseq looks for the exons by searching for capital letters within the sequence\n",
    "# exonseq = re.compile(\"[A-Z]\")\n",
    "\n",
    "# #Draw a legend of what each motif color means\n",
    "# n = 25\n",
    "# for motif,r,g,b in zip(motifs,red,green,blue):\n",
    "#     if n == 25:\n",
    "#         n = draw_legend(r,g,b,n,motif)\n",
    "#     else:\n",
    "#         n = draw_legend(r,g,b,n,motif)\n",
    "\n",
    "# ############################\n",
    "# #The way we will draw the exons will be to first draw an entire line to signify the whole gene (introns) then for each position in which\n",
    "# #an upper case character was found (exon), we will draw a vertical line. Then we will look at the motifs and draw different\n",
    "# #colored horizontal lines through each respective site\n",
    "\n",
    "\n",
    "# for lines in InputFasta:\n",
    "#     if lines.startswith(\">\"): #Begin by adding the header above each gene that will be ploted\n",
    "#         y = y + 50\n",
    "#         context.set_source_rgb(0,0,0)\n",
    "#         context.move_to(x,y-15)\n",
    "#         context.show_text(lines)\n",
    "        \n",
    "#     if not lines.startswith(\">\"): #Start by drawing our 'intron' line which is simply a line the length of our sequence\n",
    "#         context.set_line_width(1)\n",
    "#         context.move_to(x,y)\n",
    "#         context.line_to(x+len(lines),y)\n",
    "#         context.stroke()\n",
    "        \n",
    "#         #We set our line width to 10 since we will draw our motifs and exons to look pronounced\n",
    "#         context.set_line_width(10)\n",
    "        \n",
    "#         #Compite our regular expression search that looks for uppercase letters that are grouped together\n",
    "#         exonseq = re.compile(\"[A-Z]{1,}\")\n",
    "#         exonstart = []\n",
    "#         exonend = []\n",
    "        \n",
    "#         #Look through our current line and save all the exon start and stop locations\n",
    "#         for bases in exonseq.finditer(lines): #Save the location of every uppercase character in the sequence\n",
    "#             exonstart.append(bases.start()+1) #In order to account for 0 based python counts, we used start + 1\n",
    "#             exonend.append(bases.end())\n",
    "            \n",
    "#         for start, stop in zip(exonstart,exonend): #Iterate through our start and stop sites to draw our exons\n",
    "#                 draw_line(x,y,start,stop)\n",
    "        \n",
    "#         #Additionall we change our sequence lines to uppercase as the regular expressions are looking for uppercase characters\n",
    "#         lines = lines.upper()\n",
    "        \n",
    "#         #Iterate through our motif list, as well as the corresponding colors for each motif\n",
    "#         for motif,r,g,b in zip(motifs,red,green,blue):\n",
    "#             motifstart = [] #to be moved\n",
    "#             motifend = [] #to be moved\n",
    "#             motifseq = re.compile(motif,)\n",
    "        \n",
    "#             for bases in motifseq.finditer(lines):\n",
    "#                 motifstart.append(bases.start()+1) #In order to account for 0 based python counts, we used start + 1\n",
    "#                 motifend.append(bases.end())\n",
    "        \n",
    "#             for start, stop in zip(motifstart,motifend): #Iterate through our start and stop sites to draw our motifs        \n",
    "#                 context.set_source_rgb(r,g,b)\n",
    "#                 draw_line(x,y,start,stop)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.1"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
